srtcore/buffer_tools.cpp

```diff
@@ -273,5 +273,64 @@ int CSndRateEstimator::incSampleIdx(int val, int inc) const
     return val;
 }
 
+CMovingRateEstimator::CMovingRateEstimator()
+    : m_tsFirstSampleTime(sync::steady_clock::now())
+    , m_iFirstSampleIdx(0)
+    , m_iCurSampleIdx(0)
+    , m_iRateBps(0)
+    , m_Samples(NUM_PERIODS)
+{
+    resetRate(0, NUM_PERIODS);
 }
 
+void CMovingRateEstimator::addSample(int pkts, double bytes)
+{
+    const time_point now             = steady_clock::now();
+    const int        iSampleDeltaIdx = (int)count_milliseconds(now - lastSlotTimestamp) / SAMPLE_DURATION_MS;
+
+    if (iSampleDeltaIdx == 0)
+    {
+        m_Samples[m_iCurSampleIdx].m_iBytesCount += bytes;
+        m_Samples[m_iCurSampleIdx].m_iPktsCount += pkts;
+    }
+    else
+    {
+        if ((m_iCurSampleIdx + iSampleDeltaIdx) < NUM_PERIODS)
+            resetRate(m_iCurSampleIdx + 1, m_iCurSampleIdx + iSampleDeltaIdx);
+        else
+        {
+            int loopbackDiff = m_iCurSampleIdx + iSampleDeltaIdx - NUM_PERIODS;
+            resetRate(m_iCurSampleIdx + 1, NUM_PERIODS);
+            resetRate(0, loopbackDiff);
+        }
+
+        m_iCurSampleIdx                          = ((m_iCurSampleIdx + iSampleDeltaIdx) % NUM_PERIODS);
+        m_Samples[m_iCurSampleIdx].m_iBytesCount = bytes;
+        m_Samples[m_iCurSampleIdx].m_iPktsCount  = pkts;
+
+        lastSlotTimestamp = now;
+    }
+
+    computeAverageValue();
+}
+
+void CMovingRateEstimator::resetRate(int from, int to)
+{
+    for (int i = max(0, from); i < min(int(NUM_PERIODS), to); i++)
+        m_Samples[i].reset();
+}
+
+void CMovingRateEstimator::computeAverageValue()
+{
+    const time_point now           = steady_clock::now();
+    const int        startDelta    = count_milliseconds(now - m_tsFirstSampleTime);
+    const bool       isFirstPeriod = startDelta < 1000;
+    m_iRateBps                     = 0;
```

Comment:
If this isn't mutex-protected, then I think it would be better if `m_iRateBps` is an atomic variable and the value is being calculated on the intermediate local variable, then written to this field just once at the end.

Reply:
I agree, done.

Reply:
It is protected by `CUDT::m_StatsLock`

Reply:
Not exactly. This protects the statistical data from being updated while being used in order to prevent data from being read inconsistently. But this is a utility class that can be potentially used various ways. The change I proposed makes it safe no matter how it is used, and it's not at the expense of performance.

Reply:
You would need to protect the whole class then, including `m_Samples[..]`, indexes, and timestamps.

Reply:
This is only required in case when you might have statistics update coming from various threads, while I believe it's not the case. This protects the case when you have only thread feeding in the data and potentially multiple threads reading them.

Reply:
"But this is a utility class that can be potentially used various ways." And yet you assume statistics update comes from a single thread. Reading stats is protected by `CUDT::m_StatsLock`.
Putting atomics everywhere "just in case" is not the best approach.